Feat: Add path picker GUI for AnkiDroid Directory setting

ShaanNarendran opened the pull request:
<!--- Please fill the necessary details below -->
## Purpose / Description
This pr adds a path picker gui when a user wishes to select a path through the advanced settings, if the user is using the google play version, it restricts their options to their internal storage or to an external sd card (it automatically gives them the sd card option, whereas before they would have to manually find the path). If the user is on the fullrelease or full debug, then they can give permissions to manage all storage and they can manually enter their path since scoped storage wouldn't be an issue.

## Fixes
* Fixes #3114 

## Approach
David's patch ended up working very nicely and I just implemented his TO-DO's with the exception of one which I wasn't too sure about.

## How Has This Been Tested?
<img width="195" height="438" alt="image" src="https://github.com/user-attachments/assets/b1374e71-7dab-4bd8-8083-9538dcc2cf11" />
<img width="198" height="443" alt="image" src="https://github.com/user-attachments/assets/6bcc4bd7-eac4-4fc9-8b44-9fb1b91893ae" />


## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

<!--- Uncomment this section ONLY if this PR introduces new resources (external libraries, icons etc)
## Licenses
_For each new external resource, add a row to the table below:_

| Library | Description | License |
| --- | --- | --- |
| Sample Icon Library | Sample Description | [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt) |

**Maintainers:**

* [ ] Add the https://github.com/ankidroid/Anki-Android/labels/Licenses label
* [ ] Update the [licenses](https://github.com/ankidroid/Anki-Android/wiki/Licences) wiki when merging
--->

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,90 @@
+        private fun isEditText(context: Context): Boolean {
+            Permissions.canManageExternalStorage(context)
+            val prefs = androidx.preference.PreferenceManager.getDefaultSharedPreferences(context)
+            val forceScoped = prefs.getBoolean("pref_dev_force_scoped_storage_ui", false)
```
This no longer exists

ShaanNarendran committed:
Feat: Path Picker GUI

This feature allows users to choose a file path that exists on their device or an external storage device.
If the user is using a full release version, or a pre android 11 version then they have the option to edit their path to a custom one

Co-authored-by: David Allison <62114487+david-allison@users.noreply.github.com>

david-allison reviewed (changes requested):
* Select Custom Path
  * `/storage/emulated/0/Android/data/com.ichi2.anki.debug/files/AnkiDroi`
* Open Dialog Again
  * 🪲 Dialog does not show the current path as an option
  * 🪲 Selecting the top option does not apply the provided path

----

I would move the 'custom path' to a button at the bottom of the screen, with a very brief name. This means it won't be truncated at the bottom of the list if there's a number of entries

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,176 @@
+    var forceCustomSelector = false
+    var requestingCustomPath = false
+
+    private val initialLayoutResId = dialogLayoutResource
```
Add a single ine KDoc here

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,176 @@
+        return ListPreferenceTrait.Entry(label, file.absolutePath)
+    }
+
+    override var listValue: String = getPersistedString(CollectionHelper.getDefaultAnkiDroidDirectory(context).absolutePath)
```
`getDefaultAnkiDroidDirectory` can throw. This must be handled, or a user can't get out of a bad situation.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,176 @@
+            } else {
+                listOf(File(f, "AnkiDroid"))
+            }
+        }
```
This isn't good logic: multiprofile support needs to be considered, as does someone creating a custom directory on their SD Card/External Stroage via USB

david-allison reviewed (changes requested):
My main concern is that the logic to switch between `EditText` and `ListPreference` is complicated. 

Could you look into alternate solutions which require fewer variables/state?

ShaanNarendran commented:
just had a question regarding the file picker ui issue im doing. Right now, I have loadDirectories() running on the main thread and I had asked gemini to try and spot any issues and it said that having loadDirectories() on the main thread can cause ANR with devices having an older SD card and it's better to put it in the background (to me this makes sense since if the sd card can't load all the directories fast enough, the main thread is going to take too long and ANR occurs). I tried looking this up but wasn't able to find anything concrete, so I wanted to know if it's an actual issue and if its a better practice to throw it in the background?
 
 Flagging the above for discussion, I'll leave this as a TODO in the code

david-allison reviewed (changes requested):
I've asked for a second reviewer on Discord

Please ping me in 48 hours here, I'll need to give this a deeper dive, and likely submit a patch

ShaanNarendran commented:
@david-allison super sorry for the late ping, had a lot of college work to finish off but I've addressed all the comments best I could!

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    override var listEntries: List<ListPreferenceTrait.Entry> = emptyList()
+    override var listValue: String = ""
```
document that these are default values, and the real values are set later

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,190 @@
+                    .flatMap { findAnkiDroidSubDirectories(it) }
+                    .distinct()
+                    .filter { it.absolutePath != defaultDir?.absolutePath }
+                    .forEach { add(createEntry(it)) }
```
If there is 1 exception, it stops further iteration.

* Extract this to a method which is more exception-safe
* Maybe use `addAll` on the value of that method call

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,190 @@
+            val currentPath = value ?: defaultDir?.absolutePath ?: ""
+            if (currentPath.isNotEmpty() && none { it.value == currentPath }) {
+                add(0, createEntry(File(currentPath)))
+            }
```
Why is this `add(0)` rather than moving the code to the top of `buildList` (then either checking for duplicates and picking the top, or handling the duplicate check in the `add` calls)?

ShaanNarendran replied:
I implemented it like this in case of any custom paths the user had weren't found in the initial scan and that line would push it on top but I think your approach is better so ill refactor this accordingly

david-allison commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,190 @@
+        setEntryValues(listEntries.map { it.value as CharSequence }.toTypedArray())
+        listValue = value ?: defaultAnkiDir?.absolutePath ?: ""
+        setValue(listValue)
+        return FullWidthListPreferenceDialogFragment()
```
This code feels unusual - I know it's an unusual abstraction, but flagging to see if we can do better.

ShaanNarendran replied:
I think it's because there's a lot going on here and also I did have to change the entry class to allow for the mapping, so it's not the best implementation definitely

david-allison commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -473,4 +473,11 @@ this formatter is used if the bind only applies to the answer">A: %s</string>
+    <string name="pref_directory_not_set">Not set</string>
+    <string name="pref_custom_path" maxLength="41">Custom path</string>
+    <string name="pref_enter_custom_path" maxLength="41">Enter custom path</string>
+    <string name="pref_cannot_create_directory">Cannot create directory</string>
```
Is `could_not_create_dir` not usable?

david-allison reviewed (changes requested):
This looks solid!

Are tests possible here? 

Fixing the AnkiDroid directory was critical functionality back in the day, and still is on older devices.

david-allison commented:
This is one of those places which is critical to use of the app.

Feel free to add a 'copy debug info' button which copies the stack trace.

If not, more information is better here, many users here are experienced, and filesystems (SD Cards etc...) are nasty

ShaanNarendran commented:
> This is one of those places which is critical to use of the app.
> 
> Feel free to add a 'copy debug info' button which copies the stack trace.
> 
> If not, more information is better here, many users here are experienced, and filesystems (SD Cards etc...) are nasty

I think adding another button would be too much, I've appended the stack trace in the most recent push do lmk if anymore changes are needed

david-allison reviewed (changes requested):
<img width="345" height="233" alt="Screenshot 2026-01-18 at 11 22 09" src="https://github.com/user-attachments/assets/fd53829a-5c25-449d-a997-7a9a197d94f3" />

* The trailing horizontal line is unexpected
* I can no longer change my collection path, `/storage/emulated/0/Android/data/com.ichi2.anki.debug//files/19855` is a valid collection path

david-allison reviewed (changes requested):
It seems that there's a misunderstanding here, [so a test would be fantastic]

----

```
AnkiDroid Version = 2.24.0alpha2-debug (0fff3047b24701acbcaa6524fb27d6d9df95e47d)  
Backend Version = 0.1.63-anki25.09.2 (25.09.2 3890e12c9e48c028c3f12aa58cb64bd9f8895e30)  
Android Version = 15 (SDK 35)  
ProductFlavor = play  
Device Info = Google | google | emu64a | sdk_gphone64_arm64 | sdk_gphone64_arm64 | ranchu  
WebView Info = [com.google.android.webview | 749903433]: Mozilla/5.0 (Linux; Android 15; sdk_gphone64_arm64 Build/AE3A.240806.036; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/143.0.7499.34 Mobile Safari/537.36  
ACRA UUID = 3e5d97f3-05bd-45e9-9942-4c392b26bf20  
FSRS = 5.1.0 (Enabled: false)  
Crash Reports Enabled = false
```

----

* I have a standard Android [emulator] with no SD card
* I am using `play`, so we can't access standard folders on external storage (`/storage/emulated/0/AnkiDroid`)
* The one reasonable location that I can have an AnkiDroid Directory in is `Android/data/com.ichi2.anki.debug/files/`

But: Inside this folder, I can have multiple AnkiDroid Directories:

* `Android/data/com.ichi2.anki.debug/files/AnkiDroid` (default)
* `Android/data/com.ichi2.anki.debug/files/Testing` (custom)

This dialog needs to handle this case

----

In addition: selecting the root of `files` caused this to occur:

<img width="342" height="338" alt="Image" src="https://github.com/user-attachments/assets/0c5b5245-625e-4c2f-aa55-1e83f170d6c4" />

david-allison reviewed (changes requested):
This is REALLY good, one UX question:

* I create `/storage/emulated/0/AnkiDroid5`, then switch back and it's not displayed.

<img width="353" height="291" alt="Image" src="https://github.com/user-attachments/assets/ee23112f-73c0-4826-90d5-41d5e0222d41" />

Is it feasible to scan the root directory for AnkiDroid directories, or is this the best we can do?

No worries if it's not feasible, but it'd make this feature absolutely fantastic for me.

david-allison commented:
`Context.getExternalFilesDirs()` may be what you're after. Do check for both a null result, and null paths here, the API is undefined.

david-allison reviewed (changes requested):
This works great on my emulator!

I believe there are still issues with the code to get storage roots.

I wasn't able to get a test SD card working (Android Studio has now disabled the functionality)

but the path should be `storage/XXXX-XXXX`.

Here is an untested patch which I believe is correct: https://github.com/ankidroid/Anki-Android/wiki/Development-Guide#applying-a-patch

```patch
Index: AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt b/AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt	(revision fb49676afff1733fbe9909ecfb0b6fafa1b92be8)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt	(date 1772985047611)
@@ -417,3 +417,34 @@
         if (e is Error) throw e
         Result.failure(e)
     }
+
+/**
+ * Runs the provided block, catching [Exception], logging it and reporting it to [CrashReportService]
+ *
+ * **Example**
+ * ```
+ * runCatchingWithReport("callingMethod", onlyIfSilent = true) {
+ *     doSomethingRisky()
+ * }
+ * ```
+ *
+ * **Note**: This differs from [runCatching] - `Error` is thrown
+ *
+ * @param origin Data logged to Timber
+ * @param block Code to execute
+ *
+ * @throws Error If raised, this will be reported and rethrown
+ *
+ * @return A Result containing either the successful result of [block] or the [Exception] thrown
+ */
+fun <T> runCatchingWithLog(
+    origin: String? = null,
+    block: () -> T,
+): Result<T> =
+    try {
+        Result.success(block())
+    } catch (e: Throwable) {
+        Timber.w(e, origin)
+        if (e is Error) throw e
+        Result.failure(e)
+    }
Index: AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt b/AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt
--- a/AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt	(revision fb49676afff1733fbe9909ecfb0b6fafa1b92be8)
+++ b/AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt	(date 1772985438962)
@@ -19,7 +19,6 @@
 
 import android.content.Context
 import android.graphics.Color
-import android.os.Environment
 import android.text.Spannable
 import android.text.SpannableString
 import android.text.style.ForegroundColorSpan
@@ -30,6 +29,7 @@
 import androidx.preference.ListPreferenceDialogFragmentCompat
 import com.ichi2.anki.CollectionHelper
 import com.ichi2.anki.R
+import com.ichi2.anki.runCatchingWithLog
 import com.ichi2.anki.showThemedToast
 import com.ichi2.utils.input
 import com.ichi2.utils.negativeButton
@@ -88,11 +88,16 @@
             .map(::absolutePathToDisplayEntry)
 
     private fun isValidAnkiDir(dir: File): Boolean {
-        if (!dir.isDirectory || dir.name.startsWith(".")) return false
-        if (IGNORED_DIRECTORIES.contains(dir.name.lowercase())) return false
-        if (dir.name.startsWith("AnkiDroid", ignoreCase = true)) return true
-        val contents = dir.list() ?: return false
-        return contents.any { it == "collection.anki2" }
+        try {
+            if (!dir.isDirectory || dir.name.startsWith(".")) return false
+            if (IGNORED_DIRECTORIES.contains(dir.name.lowercase())) return false
+            if (dir.name.startsWith("AnkiDroid", ignoreCase = true)) return true
+            val contents = dir.list() ?: return false
+            return contents.any { it == "collection.anki2" }
+        } catch (e: Exception) {
+            Timber.w(e, "Could not access directory")
+            return false
+        }
     }
 
     /**
@@ -100,41 +105,28 @@
      * If one directory fails to scan, we log it and continue to the next one
      */
     private fun getScannedDirectories(): List<File> {
-        // Get all possible storage roots
-        val roots = mutableListOf<File>()
-        try {
-            val appDirs = context.getExternalFilesDirs(null) ?: emptyArray()
-            for (dir in appDirs) {
-                if (dir == null) continue
-                val path = dir.absolutePath
-                val androidIndex = path.indexOf("/Android/")
-                if (androidIndex != -1) {
-                    roots.add(File(path.take(androidIndex)))
-                }
+        val storageRoots = runCatchingWithLog("get storage roots") {
+                context.getExternalFilesDirs(null)
             }
-        } catch (e: Exception) {
-            Timber.w(e, "Critical error getting storage roots")
-            return emptyList()
+            .getOrNull()
+            .orEmpty()
+            .filterNotNull()
+
+        fun File.getValidAnkiSubfolders(): List<File>? {
+            val subFolders = runCatchingWithLog("Could not list files") {
+                listFiles()
+            }.getOrNull() ?: return null
+
+            return subFolders.filter { isValidAnkiDir(it) }.ifEmpty { null }
         }
-        val candidates = mutableListOf<File>()
-        for (root in roots) {
-            // Find the subfolders of each root and their respective valid directories (containing collection.anki2
-            val subFolders =
-                try {
-                    root.listFiles { f -> f.isDirectory && !f.name.startsWith(".") }
-                } catch (e: Exception) {
-                    Timber.w(e, "Could not list files in $root")
-                    null
-                }
-            val validChildren = subFolders?.filter { isValidAnkiDir(it) } ?: emptyList()
-            if (validChildren.isNotEmpty()) {
-                candidates.addAll(validChildren)
-            } else {
+
+        return storageRoots.flatMap { root ->
+            when (val ankiFolders = root.getValidAnkiSubfolders()) {
                 // If no anki directories are found, we can list this as it is likely an SD card
-                candidates.add(root)
+                null -> listOf(root)
+                else -> ankiFolders
             }
-        }
-        return candidates.distinct()
+        }.distinct()
     }
 
     // TODO: Possibly move loadDirectories() to a background thread if ANR occurs
```

----

It would be really useful to future implementers if you added a singular test in `androidTest`

Just with the scaffolding code to set up, display the dialog, and see the list results.

ShaanNarendran commented on AnkiDroid/src/main/java/com/ichi2/preferences/ExternalDirectorySelectionPreference.kt:
```diff
@@ -0,0 +1,227 @@
+        } catch (e: Exception) {
+            Timber.w(e, "Critical error getting storage roots")
+            return emptyList()
+        }
```
Do we not want to allow users to make custom paths out of the android directory? if so I can remove the map function I added

ShaanNarendran commented:
<details> 
<summary>Patch</summary>
 
```patch
Index: AnkiDroid/src/androidTest/java/com/ichi2/anki/preferences/ExternalDirectorySelectionPreferenceTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/androidTest/java/com/ichi2/anki/preferences/ExternalDirectorySelectionPreferenceTest.kt b/AnkiDroid/src/androidTest/java/com/ichi2/anki/preferences/ExternalDirectorySelectionPreferenceTest.kt
new file mode 100644
--- /dev/null	(date 1773519838146)
+++ b/AnkiDroid/src/androidTest/java/com/ichi2/anki/preferences/ExternalDirectorySelectionPreferenceTest.kt	(date 1773519838146)
@@ -0,0 +1,77 @@
+package com.ichi2.anki.preferences
+/*
+ Copyright (c) 2026 Shaan Narendran <shaannaren06@gmail.com>
+
+ This program is free software; you can redistribute it and/or modify it under
+ the terms of the GNU General Public License as published by the Free Software
+ Foundation; either version 3 of the License, or (at your option) any later
+ version.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License along with
+ this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import android.os.Bundle
+import androidx.fragment.app.testing.launchFragmentInContainer
+import androidx.preference.PreferenceFragmentCompat
+import androidx.preference.PreferenceScreen
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.RootMatchers.isDialog
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.withText
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.ichi2.anki.R
+import com.ichi2.anki.testutil.GrantStoragePermission
+import com.ichi2.anki.testutil.grantPermissions
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import com.ichi2.preferences.ExternalDirectorySelectionPreference
+
+@RunWith(AndroidJUnit4::class)
+class ExternalDirectorySelectionPreferenceTest {
+
+    @get:Rule
+    val runtimePermissionRule = grantPermissions(GrantStoragePermission.storagePermission)
+
+    class TestPreferenceFragment : PreferenceFragmentCompat() {
+        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
+            val context = preferenceManager.context
+            val screen: PreferenceScreen = preferenceManager.createPreferenceScreen(context)
+            val externalDirPref = ExternalDirectorySelectionPreference(context, null).apply {
+                key = "test_external_dir_pref"
+                title = "AnkiDroid directory"
+            }
+            screen.addPreference(externalDirPref)
+            preferenceScreen = screen
+        }
+
+        @Suppress("DEPRECATION")
+        override fun onDisplayPreferenceDialog(preference: androidx.preference.Preference) {
+            if (preference is ExternalDirectorySelectionPreference) {
+                val dialog = preference.makeDialogFragment()
+                dialog.arguments = Bundle().apply { putString("key", preference.key) }
+                dialog.setTargetFragment(this, 0)
+                dialog.show(parentFragmentManager, "androidx.preference.PreferenceFragment.DIALOG")
+            } else {
+                super.onDisplayPreferenceDialog(preference)
+            }
+        }
+    }
+
+    @Test
+    fun testDialogDisplaysScannedDirectories() {
+        val scenario = launchFragmentInContainer<TestPreferenceFragment>(themeResId = androidx.appcompat.R.style.Theme_AppCompat)
+        scenario.onFragment { fragment ->
+            val pref = fragment.findPreference<ExternalDirectorySelectionPreference>("test_external_dir_pref")!!
+            fragment.onDisplayPreferenceDialog(pref)
+        }
+        onView(withText(R.string.pref_custom_path))
+            .inRoot(isDialog())
+            .check(matches(isDisplayed()))
+    }
+}
\ No newline at end of file
```
   
</details>

david-allison reviewed (changes requested):
Almost there. One question on `/Android/` please provide examples in the code

ShaanNarendran commented:
> I do have a spare phone which supports it, but it'd be ideal if you could test this with an emulator

I think I tried in the very beginning but I wasn't able to figure out how to set it up, ill try again now after making the final changes

david-allison reviewed (approved):
On my API 28 Huawei LLD-31, this allowed me to select my SD card, without needing to figure out what the path was. AMAZING JOB!!!!

This does so much good for our users, especially ones who are running on older phones and NEED an SD card to keep it going. Well done!!!!

<img width="361" height="723" alt="Image" src="https://github.com/user-attachments/assets/344000d3-d37a-496f-9ea1-51f421ae2d46" />

BrayanDSO commented:
so, how do I remove a path?

ShaanNarendran commented:
> so, how do I remove a path?

There's no functionality there to do so, the main use case for this feature would be so that users can manually input/find paths for external storage like sd card or maybe input an external path that they want to use. I think I could add a remove button but maybe it's too much work for the few who would try to abuse this by adding too many paths? It's just an improvement to what we had before.

BrayanDSO commented:
The use cases are 1. mispelling, 2. giving up on having 2 or more "profiles", 3. Testing out stuff then deleting the collection

ShaanNarendran commented:
> The use cases are 1. mispelling, 2. giving up on having 2 or more "profiles", 3. Testing out stuff then deleting the collection

I can add this, but isn't it dangerous to allow the user to delete paths that might contain their collections? Maybe, they could hide that path or just delete the string so then they'd have to re-enter but then that file would still be in their phone which would cause bloat. Is there a good way to handle this?

criticalAY commented:
I will be taking over this one, took permission from OP
